### PR TITLE
feat(secrecy): enable data secrecy for organizations

### PR DESCRIPTION
**Pull request by cathteng**

### CODE REVIEW

File: src/sentry/api/exceptions.py [hunk 1]
```diff
@@ -85,6 +85,11 @@ class SuperuserRequired(SentryAPIException):
     message = "You need to re-authenticate for superuser."
 
 
+class DataSecrecy(SentryAPIException):
```
Comment (evanpurkhiser) [inline src/sentry/api/exceptions.py:88]: Should this end in “Exception”? Can we add a comment describing what this is for?

File: src/sentry/conf/server.py [hunk 2]
```diff
@@ -1354,6 +1354,8 @@ def SOCIAL_AUTH_DEFAULT_USERNAME() -> str:
     "organizations:discover-basic": True,
     # Enable discover 2 custom queries and saved queries
     "organizations:discover-query": True,
+    # Enables data secrecy mode
```
Comment (evanpurkhiser) [inline src/sentry/conf/server.py:1357]: Can we expand on this more? What is data secrecy mode. Imagine you’re a user scrolling g through this list and you were like “what is that”

File: src/sentry/api/exceptions.py [hunk 3]
```diff
@@ -85,6 +85,11 @@ class SuperuserRequired(SentryAPIException):
     message = "You need to re-authenticate for superuser."
 
 
+class DataSecrecyError(SentryAPIException):
+    status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
```
Comment (dashed) [inline src/sentry/api/exceptions.py:89]: @cathteng I don't think 5xx error isn't appropriate for this type of exception. 

403 Forbidden might be better: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403

File: src/sentry/api/permissions.py [hunk 4]
```diff
@@ -122,6 +124,13 @@ def determine_access(
         if org_context is None:
             assert False, "Failed to fetch organization in determine_access"
 
+        if (
+            request.user
+            and request.user.is_superuser
+            and features.has("organizations:enterprise-data-secrecy", org_context.organization)
+        ):
+            raise DataSecrecyError()
```
Comment (dashed) [inline src/sentry/api/permissions.py:132]: @cathteng it'll be nice to have a test case for these conditions.

File: src/sentry/api/exceptions.py (hunk 3 above)
Comment (cathteng) [reply src/sentry/api/exceptions.py:89]: if we return 403, we will be prompting the user to login as superuser https://github.com/getsentry/sentry/blob/master/static/app/views/organizationContextContainer.tsx#L245-L255

**Review by dashed — APPROVED** [review]

File: src/sentry/api/exceptions.py [hunk 5]
```diff
@@ -85,6 +85,11 @@ class SuperuserRequired(SentryAPIException):
     message = "You need to re-authenticate for superuser."
 
 
+class DataSecrecyError(SentryAPIException):
+    status_code = status.HTTP_418_IM_A_TEAPOT
```
Comment (dcramer) [inline src/sentry/api/exceptions.py:89]: shouldnt this be a 401? or 403?

**Review by dashed — APPROVED** [review]

**Review by leedongwei — APPROVED** [review]

**Pull request merged**